flyteplugins/go/tasks/pluginmachinery/ioutils/remote_file_output_reader.go

```diff
@@ -3,75 +3,243 @@ package ioutils
+	if len(pieces) != 2 {
+		err = errors.Errorf("expected canonical error filename to have a single dot (.), got %d", len(pieces))
+		return
+	}
+	errorFilePrefix := pieces[0]
+	scheme, container, key, _ := e.errorDirPath.Split()
+	errorFilePathPrefix = storage.NewDataReference(scheme, container, filepath.Join(key, errorFilePrefix))
+	errorFileExtension = fmt.Sprintf(".%s", pieces[1])
+	return
+}
+
+func (e *EarliestFileErrorRetriever) HasError(ctx context.Context) (bool, error) {
+	errorFilePathPrefix, errorFileExtension, err := e.parseErrorFilename()
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to parse canonical error filename @[%s]", e.canonicalErrorFilename)
+	}
+	hasError := false
+	const maxItems = 1000
+	cursor := storage.NewCursorAtStart()
+	for cursor != storage.NewCursorAtEnd() {
+		var err error
+		var errorFilePaths []storage.DataReference
+		errorFilePaths, cursor, err = e.store.List(ctx, errorFilePathPrefix, maxItems, cursor)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to list error files @[%s]", e.errorDirPath)
+		}
+		for _, errorFilePath := range errorFilePaths {
+			if strings.HasSuffix(errorFilePath.String(), errorFileExtension) {
+				metadata, err := e.store.Head(ctx, errorFilePath)
+				if err != nil {
+					return false, errors.Wrapf(err, "failed to read error file @[%s]", errorFilePath)
+				}
+				err = e.validatePayloadSize(metadata)
+				if err != nil {
+					return false, errors.Wrapf(err, "error file @[%s] failed payload size validation", errorFilePath)
+				}
+				hasError = true
+			}
+		}
+	}
+	return hasError, nil
+}
+
+func (e *EarliestFileErrorRetriever) GetError(ctx context.Context) (io.ExecutionError, error) {
+	errorFilePathPrefix, errorFileExtension, err := e.parseErrorFilename()
+	if err != nil {
+		return io.ExecutionError{}, errors.Wrapf(err, "failed to parse canonical error filename @[%s]", e.canonicalErrorFilename)
+	}
+	const maxItems = 1000
+	cursor := storage.NewCursorAtStart()
+	type ErrorFileAndDocument struct {
+		errorFilePath storage.DataReference
+		errorDoc      *core.ErrorDocument
+	}
+	var errorFileAndDocs []ErrorFileAndDocument
+	for cursor != storage.NewCursorAtEnd() {
+		var err error
+		var errorFilePaths []storage.DataReference
+		errorFilePaths, cursor, err = e.store.List(ctx, errorFilePathPrefix, maxItems, cursor)
+		if err != nil {
+			return io.ExecutionError{}, errors.Wrapf(err, "failed to list error files @[%s]", e.errorDirPath)
+		}
+		for _, errorFilePath := range errorFilePaths {
+			if strings.HasSuffix(errorFilePath.String(), errorFileExtension) {
+				errorDoc := &core.ErrorDocument{}
+				err := e.store.ReadProtobuf(ctx, errorFilePath, errorDoc)
+				if err != nil {
+					return io.ExecutionError{}, errors.Wrapf(err, "failed to read error file @[%s]", errorFilePath.String())
+				}
+				errorFileAndDocs = append(errorFileAndDocs, ErrorFileAndDocument{errorFilePath: errorFilePath, errorDoc: errorDoc})
```

Comment:
Instead of appending all of them here, can we just keep the oldest around... so by the time the loop is done you have a single document?

Reply:
Sure. Will update.

Reply:
Addressed.

Reply:
?

Reply:
@wild-endeavor Rather than collecting them to a list and picking from the list the one with the smallest timestamp, we now do it in one loop.